url73e3057d61_empeeoh_BACnet4J_tgz-pJ8-com_serotonin_bacnet4j_test_OverflowTestJ8/src/com/serotonin/bacnet4j/service/confirmed/AtomicWriteFileRequest.java

```diff
@@ -0,0 +1,214 @@
+import com.serotonin.bacnet4j.type.primitive.SignedInteger;
+import com.serotonin.bacnet4j.type.primitive.UnsignedInteger;
+import org.free.bacnet4j.util.ByteQueue;
+
+public class AtomicWriteFileRequest extends ConfirmedRequestService {
+    private static final long serialVersionUID = 2426317127743066428L;
+
+    public static final byte TYPE_ID = 7;
+
+    private final ObjectIdentifier fileIdentifier;
+    private SignedInteger fileStart;
+    private OctetString fileData;
+    private UnsignedInteger recordCount;
+    private SequenceOf<OctetString> fileRecordData;
+
+    public AtomicWriteFileRequest(ObjectIdentifier fileIdentifier, SignedInteger fileStart, OctetString fileData) {
+        super();
+        this.fileIdentifier = fileIdentifier;
+        this.fileStart = fileStart;
+        this.fileData = fileData;
+    }
+
+    public AtomicWriteFileRequest(ObjectIdentifier fileIdentifier, SignedInteger fileStart,
+            UnsignedInteger recordCount, SequenceOf<OctetString> fileRecordData) {
+        super();
+        this.fileIdentifier = fileIdentifier;
+        this.fileStart = fileStart;
+        this.recordCount = recordCount;
+        this.fileRecordData = fileRecordData;
+    }
+
+    @Override
+    public byte getChoiceId() {
+        return TYPE_ID;
+    }
+
+    @Override
+    public AcknowledgementService handle(LocalDevice localDevice, Address from, OctetString linkService)
+            throws BACnetException {
+        AtomicWriteFileAck response;
+
+        BACnetObject obj;
+        FileObject file;
+        try {
+            // Find the file.
+            obj = localDevice.getObjectRequired(fileIdentifier);
+            if (!(obj instanceof FileObject))
+                throw new BACnetServiceException(ErrorClass.object, ErrorCode.rejectInconsistentParameters);
+            file = (FileObject) obj;
+
+            // Validation.
+            FileAccessMethod fileAccessMethod = (FileAccessMethod) file
+                    .getProperty(PropertyIdentifier.fileAccessMethod);
+            if (fileData == null && fileAccessMethod.equals(FileAccessMethod.streamAccess) || fileData != null
+                    && fileAccessMethod.equals(FileAccessMethod.recordAccess))
+                throw new BACnetErrorException(getChoiceId(), ErrorClass.object, ErrorCode.invalidFileAccessMethod);
+        }
+        catch (BACnetServiceException e) {
+            throw new BACnetErrorException(getChoiceId(), e);
+        }
+
+        if (fileData == null) {
+            throw new NotImplementedException();
+        }
+
+        long start = fileStart.longValue();
+
+        if (start > file.length())
+            throw new BACnetErrorException(getChoiceId(), ErrorClass.object, ErrorCode.invalidFileStartPosition);
+
+        try {
+            file.writeData(start, fileData);
+            response = new AtomicWriteFileAck(fileData == null, fileStart);
+        }
+        catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)